src/services/navigationBar.ts

```diff
@@ -103,138 +103,144 @@ namespace ts.NavigationBar {
         parent = parentsStack.pop();
     }
 
-    function addNodeWithRecursiveChild(node: Node, child: Node): void {
+    function addNodeWithRecursiveChild(node: Node, child: Node, addChildrenRecursively: (node: Node) => void): void {
         startNode(node);
         addChildrenRecursively(child);
         endNode();
     }
 
     /** Look for navigation bar items in node's subtree, adding them to the current `parent`. */
-    function addChildrenRecursively(node: Node): void {
-        if (!node || isToken(node)) {
-            return;
-        }
+    function addChildrenRecursively(node: Node, cancellationToken: CancellationToken): void {
+        function addChildrenRecursively(node: Node): void {
+            cancellationToken.throwIfCancellationRequested();

```

Comment:
Check the impact of this. If it's called to frequently, you might want to throttle the checks based on count/time.